Fixes #10482 - RewriteHandler with multiple HeaderPatternRules.

sbordet opened the pull request:
Reviewed the implementation of RewriteHandler, that was broken for those rules that were overriding `Rule.Handler.handle()`.

The problem was that the handling was not forwarded along the chain of rules, so only the last one was applied.

Removed (a breaking change) `Rule.Handler.handle(Response, Callback)` because it is necessary that the outermost request wrapper is passed to the `RewriteHandler` child `Handler`. Updated implementations to override `handle(Request, Response, Callback)` instead, so that the request parameter is preserved along the chain.

sbordet committed:
Fixes #10482 - RewriteHandler with multiple HeaderPatternRules.

Reviewed the implementation of RewriteHandler, that was broken for those rules that were overriding `Rule.Handler.handle()`.

The problem was that the handling was not forwarded along the chain of rules, so only the last one was applied.

Removed (a breaking change) `Rule.Handler.handle(Response, Callback)` because it is necessary that the outermost request wrapper is passed to the `RewriteHandler` child `Handler`.
Updated implementations to override `handle(Request, Response, Callback)` instead, so that the request parameter is preserved along the chain.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented:
@sbordet As per our call, I don't like that each `Rule.Handler` is-a `Request` and is passed a `Request` in `handle`.

I think we removed the request parameter to avoid that.   Consider the following example to how I imagine it was meant to be working.

 + We have a structure of  `Server -> RewriteHandler -> MyAppHandler`
 + The `RewriteHandler` has 3 rules A, B & C, all of which modify the request in some way by wrapping it.
 + Ultimately that means that `MyAppHandler` should be called with a request that is `c{b{a{request}}}`
 + But each rule also produces creates a `Rule.Handler` instance and those handlers should be invoked in order: `a->b->c->MyAppHandler`
 + Moreover, each `Rule.Handler` should see the request at the appropriate level of wrapping, thus the execution order should be:
    1. `A.handle(request, response, callback)`
    2. `B.handle(a{request}, response, callback)`
    3. `C.handle(b{a{request}}, response, callback)`
    4. `MyAppHandler.handle(c{b{a{request}}}, response, callback)`

The order of the invocation of the handlers could be important.   Let's say that all the handle methods of all the rules did a `response.getHeaders().put("X-Ruled", someSpecialValue)`, then once we get to `MyAppHandler`, the value of that header should be the value that C set, as it overwrote B, which overwrite A.

Moreover, each Rule.Handler should see the request only at the level of wrapping that was applied at the time it was matched.    Consider if rule `A.handle` did `response.getHeaders().put("X-A-cause", request.getHttpURI())`, then it needs to see the request as only `a{request}` so it gets the correct URI.  If it sees `c{b{a{request}}}`, then it will set the wrong URI.

So I think that each `Rule.Handle` should have a final `handle` method that does something like:
```java
    public final boolean handle(Request request, Response response, Callback callback)
    {
        Request wrapped = getWrapped();
        assert wrapped.getRequest() == request;

        return doHandle(wrapped, response, callback);
    }

    // Override this method to handle the rule.
    protected boolean doHandle(Request request, Response response, Callback callback)
    {
        assert getRequest() == request;
        getHandler().handle(request, response, callback);
    }
```

Edit: extra thoughts below

To achieve this, each `Rule.matchAndApply` needs to do something like:

```java
    public Handler matchAndApply(Request request, Iterator<Rule> nextRules, Handler ultimateHandler) throws IOException
    {
         if (match(input))
         {
             Request wrapper = wrap(request);
             Rule.Handler handler = new RuleHandle(wrapper);

             handler.setHandler(nextRules.hasNext()
                 ? nextRules.next().matchAndApply(wrapper, nextRules))
                 : ultimateHandler;
             return  handler 
         }
   
         return nextRules.hasNext()
                 ? nextRules.next().matchAndApply(request, nextRules))
                 : ultimateHandler;
    }
```

sbordet committed:
Fixes #10482 - RewriteHandler with multiple HeaderPatternRules.

Updated the implementation after review.

Restored the `Rule.Handler.handle(Response, Callback)` method so that existing code won't break.

Now the wrapping at the constructor produces RH3(RH2(RH1(Request))), but the handling is performed from the innermost towards the outermost.
In this way, the order of rules is respected, both in the wrapping at rule application, and in the `Rule.Handler` handling.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@gregw I went a hopefully simpler path, by double linking the `RuleHandler` so that rule applications build wrappers as in `c{b{a{request}}}`, but calling `handle()` runs in the opposite direction so `{a.handle() -> b.handle() -> c.handle() -> MyAppHandler.handle()}`.

There is no need for a `Rule.Handler` to implement `Request.Handler` now, because we never call that method (so does not need to be made `final`), so I removed that to simplify.

gregw commented on jetty-core/jetty-rewrite/src/main/java/org/eclipse/jetty/rewrite/handler/Rule.java:
```diff
@@ -29,94 +29,96 @@ public abstract class Rule
+    public static class Handler extends Request.Wrapper
     {
-        private volatile Handler _handler;
+        private Rule.Handler _wrapper;
```
```suggestion
        private Rule.Handler _nextRuleHandler;
```

gregw commented on jetty-core/jetty-rewrite/src/main/java/org/eclipse/jetty/rewrite/handler/RewriteHandler.java:
```diff
@@ -123,14 +131,32 @@ public boolean handle(Request request, Response response, Callback callback) thr
 
         // No rule matched, call super with the original request.
-        if (output == null)
+        if (result == null)
```
Maybe
```suggestion
        if (result == null || result == input)
```

gregw commented on jetty-core/jetty-rewrite/src/main/java/org/eclipse/jetty/rewrite/handler/RewriteHandler.java:
```diff
@@ -123,14 +131,32 @@ public boolean handle(Request request, Response response, Callback callback) thr
-        return output.handle(output, response, callback);
+        // At least one rule matched, link the last Rule.Handler
+        // to invoke the child Handler of this RewriteHandler.
+        new Output(result, getHandler());
```
```suggestion
        new LastRuleHandler(result, getHandler());
```

gregw replied:
Also it would be clearer to explicitly link here rather than use side effect within constructor

gregw commented on jetty-core/jetty-rewrite/src/main/java/org/eclipse/jetty/rewrite/handler/RewriteHandler.java:
```diff
@@ -123,14 +131,32 @@ public boolean handle(Request request, Response response, Callback callback) thr
+        return input.handle(response, callback);
+    }
+
+    private static class Output extends Rule.Handler
```
```suggestion
    private static class LastRuleHandler extends Rule.Handler
```

gregw reviewed (approved):
That's simpler.... but maybe some renames for clarity.  Output is no longer Output

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
